File: python/queens-that-can-attack-the-king.py
DIRECTIONS = [
    (1, 1), (-1, -1), (1, -1), (-1, 1),
    (0, 1), (0, -1), (1, 0), (-1, 0),
]

LIMIT_X, LIMIT_Y = 8, 8

class Solution:
    def is_king_attacked(queens: List[List[int]], king: List[int], x: int, y: int) -> (int, int):
        current_x, current_y = king[0] + x, king[1] + y
        while (0 <= current_x < LIMIT_X) and (0 <= current_y < LIMIT_Y):
            if (current_x, current_y) in queens:
                return (current_x, current_y)
            current_x += x
            current_y += y
        return None

    def queensAttacktheKing(self, queens: List[List[int]], king: List[int]) -> List[List[int]]:
        queens_count = []
        queens = {(x, y) for x, y in queens}
        for x, y in DIRECTIONS:
            queen = Solution.is_king_attacked(queens, king, x, y)
            if queen != None:
                queens_count.append(queen)
        return queens_count

File: go/queens-that-can-attack-the-king.go
// Declare a global 2D array
var DIRECTIONS = [8][2]int{
	{0, 1}, {0, -1}, {1, 0}, {-1, 0},
	{1, 1}, {1, -1}, {-1, 1}, {-1, -1},
}

var BOARD_SIZE = 8

func BuildQueenSet(queens [][]int) map[int]bool {
	queenSet := make(map[int]bool)
	for i := 0; i < len(queens); i++ {
		position := (10*queens[i][0] + queens[i][1])
		queenSet[position] = true
	}
	return queenSet
}

func IsKingAttacked(queenSet map[int]bool, king []int, direction [2]int) ([]int, bool) {
	x := king[0] + direction[0]
	y := king[1] + direction[1]
	isKingAttacked := false
	queen := []int{-1, -1}
	for 0 <= x && x < BOARD_SIZE && 0 <= y && y < BOARD_SIZE {
		position := 10*x + y
		if queenSet[position] {
			isKingAttacked = true
			queen[0] = x
			queen[1] = y
			break
		}
		x += direction[0]
		y += direction[1]
	}
	return queen, isKingAttacked
}

func queensAttacktheKing(queens [][]int, king []int) [][]int {
	queenSet := BuildQueenSet(queens)
	queensAttacking := make([][]int, 0)
	for i := 0; i < len(DIRECTIONS); i++ {
		queen, isKingAttacked := IsKingAttacked(queenSet, king, DIRECTIONS[i])
		if isKingAttacked {
			queensAttacking = append(queensAttacking, queen)
		}
	}
	return queensAttacking
}
